src/systems/filecoin_markets/retrieval_market/_index.md

```diff
@@ -1,23 +1,75 @@
+- A provider module to respond to queries and deal proposals
 
-# Retrieval V0 `libp2p` Services
+# VO & V1
 
-The v0 `retrieval market` will initially be implemented as two `libp2p` services. It will be request response based, where the client who is requesting a file sends a `retrieval deal proposal` to the miner. The miner chooses whether or not to accept it, sends their response which (if they accept the proposal) includes a `signed retrieval deal`, followed by the actual requested content, streamed as a series of bitswap block messages, using a pre-order traversal of the dag. Each block should use the [bitswap block message format](https://github.com/ipfs/go-bitswap/blob/c980d7ed36f278e93828acf920f3a911e8263265/message/message.go#L228). This way, the client should be able to verify the data incrementally as it receives it. Once the client has received all the data, it should then send a payment channel SpendVoucher of the proposed amount to the miner. This protocol may be easily extended to include payments from the client to the miner every N blocks, but for now we omit that feature.
+The retrieval markets will evolve to support sending arbitrary payload CID's & selectors within a piece (V1). The first version (V0) will only support fetching the payload CID which is at the root of PieceCID's Car File, and will only support fetching the whole DAG.
+
+The first version will send data over the retrieval protocol itself a series of Blocks encoded in Bitswap format and verify received blocks manually. 
+
+The protocol will evolve to piggy back on the Data Transfer system and Graphsync to handle transfer and verification, and to support arbitrary selectors, and to reduce roundtrips. The Data Transfer System will be augmented to support pausing/resuming and sending intermediate vouchers to facilitate this
+
+The underlying protocols will change, but the API interfaces for the client & provider will not change
+
+# Deal Flow (V0)
+
+{{< diagram src="retrieval_flow_v0.mmd.svg" title="Retrieval Flow - V0" >}}
+
+The baseline version of proposing and accepting a deal will work as follows:
+
+- The client finds a provider of a given piece with FindProviders.
+- The client queries a provider to see if it meets its retrieval criteria (via Query Protocol)
+- The client creates a payment channel as neccesary and a lane, ensures there are free funds in the channel
+- The client sends a RetrievalDealProposal to the retrieval miner. (via RetrievalProtocol)
+- The providers validates the proposal and rejects it if it is invalid: invalid price per byte, invalid payment channel are reasons for rejection
+- If the request is valid, the provider responds to with an accept message
+- The provider may also send a small number of blocks over the protocol before requiring payment
+- The client consumes blocks over the retrieval protocol and manually verifies them
+- When the requires payment to proceed, it sends a request for payment and does not send any more blocks
```

Comment:
should the rate of payment requests be automated as part of the retrieval deal to prevent a de facto DoS from a hungry provider?

Reply:
Likewise, should there be a timeout after which if either party has not acted (paid, sent bytes), the deal is terminated and they are penalized?